validator/client/src/test/java/tech/pegasys/teku/validator/client/StableSubnetSubscriberTest.java

```diff
@@ -0,0 +1,108 @@
+/*
+ * Copyright 2020 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package tech.pegasys.teku.validator.client;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.ArgumentMatchers.argThat;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.verifyNoMoreInteractions;
+
+import com.google.common.primitives.UnsignedLong;
+import java.util.Set;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.mockito.ArgumentCaptor;
+import tech.pegasys.teku.validator.api.SubnetSubscription;
+import tech.pegasys.teku.validator.api.ValidatorApiChannel;
+
+public class StableSubnetSubscriberTest {
```

Comment:
We should add a test to cover when the number of validators means we should subscribe to more subnets than exist. And for decreasing the number of validators when we still have enough to stay subscribed to all the subnets.

Also for decreasing the number of validators to 0.

Reply:
I'd also suggest injecting the `Random` instance into `StableSubnetSubscriber` then you could inject one with a known seed (or a mock) and know which subscriptions should be added or removed.

Reply:
Added all the tests you've suggested.

Reply:
I do inject the Random object now, but I think it would be really complex to be able to mock which subscriptions should be added or removed, because we store the subnet's in a set, and use streams to pick one random object from set.

Reply:
For the `shouldReplaceExpiredSubscriptionsWithNewOnes` test, I decided to take your advice and try mocking the random object. It worked well for calculating the subscription slot in a visible fashion, but about the subnet ids, I noticed that the `availableSubnetIndices`set is keeping them in order. I don't think my test should be depending on a Set keeping elements in order since I don't think its part of Set API to be kept in order. However, I'll keep it like it is for now because I feel like you might have another insight.

Reply:
Reverted to not injecting the `Random` instance in light of the selection of subnets from a set stream.